### DISCUSSION TITLE
About Korean ASR

### DISCUSSION THREAD

**Discussion by nycsv** [NVIDIA-NeMo/NeMo#3648] [Public Models]
Hi guys,

Thank you for sharing a great tool for conversational AI.

I'm going to start a discussion on Korean ASR here. @okuchaiev

**Comment by nycsv** [comment]
First, I share one of the models in the following.


### Conformer-Transducer-BPE-Small.nemo [[Link](https://drive.google.com/file/d/1sAGidT3_UbKb0uSoUzc5Lh24zqwDBREj/view?usp=sharing)]

## **Model Overview**

This collection contains small size versions of Conformer-Transducer trained on ksponspeech which is an open-domain Korean dialog corpus.

## **Model Architecture**

Conformer-Transducer model is an autoregressive variant of Conformer model [1] for Automatic Speech Recognition which uses Transducer loss/decoding. You may find more info on the detail of this model here: [[Conformer-Transducer Model](https://docs.nvidia.com/deeplearning/nemo/user-guide/docs/en/main/asr/models.html)].

## **Training**

The NeMo toolkit [3] was used for training the models for over several hundred epochs. These model are trained with this [[base config](https://github.com/NVIDIA/NeMo/blob/main/examples/asr/conf/conformer/conformer_transducer_bpe.yaml)].

The tokenizers for these models were built using the text [transcript](https://github.com/NVIDIA/NeMo/blob/main/scripts/tokenizers/process_asr_text_tokenizer.py).

## Datasets

All the models in this collection are trained on Ksponspeech dataset [[Download](https://aihub.or.kr/aidata/105/download)]

## **Performance**

The list of the available models in this collection is shown in the following. Performances of the ASR models are reported in terms of Word Error Rate (WER%) with mAES decoding.


Version | Tokenizer | eval_clean CER | eval_other CER | eval_clean WER | eval_other WER
--- | --- | --- | --- |--- |---
v1.5.1 | SentencePiece Char | 7.97% | 8.85% | 20.94% | 25.18%

**Reply by titu1994** [reply]
This is awesome ! 1000 hours of Korean speech annotated with high quality is really good source for training ! Also, this is a small model with 10 M params only, I'm pretty sure the medium or large model will perform even better, so thank you for this awesome checkpoint !

If you don't mind, could you include greedy decoding scores in the table ?

**Reply by nycsv** [reply]
Yeah I will share a medium model soon. I hope it performs better! 

These are the performances with greedy decoding.

Version | Tokenizer | eval_clean CER | eval_other CER | eval_clean WER | eval_other WER
--- | --- | --- | --- |--- |---
v1.5.1 | SentencePiece Char | 8.12% | 8.62% | 21.41% | 24.79%

**Reply by titu1994** [reply]
Quite strong scores even with greedy !

**Reply by okuchaiev** [reply]
Did you start with pre-trained English encoder? (We found that this usually helps, especially if you have limited compute and/or data)

**Reply by nycsv** [reply]
The model was trained from scratch. I firstly approached by fine-tuning the English-based pretrained model for Korean model, but I found that this was not different from not using the pretrained model by checking the loss trend in the first few epochs. And I didn't progress further.

I wonder if there was an experience of relatively improving the performance by using the english-based pretrained model for a non-alphabet character (e.g. chinese, japanese)-based ASR model.

**Reply by titu1994** [reply]
Pretrained checkpoints will almost always end up converting to better results, even if the loss curves look similar. For example the difference in Mandarin of from scratch and finetuned English is roughly 1% CER absolute. https://arxiv.org/abs/2104.01721 table 6. And that's an old model - now the current model has CER 5.2% so that's down from 6.2%.

**Reply by nycsv** [reply]
Awesome! Thank you for sharing. I will start to train new model with English pretrained model.

**Reply by nycsv** [reply]
Hi @okuchaiev @titu1994, 

The performances have been updated as follows,

Version | Tokenizer | eval_clean CER | eval_other CER | eval_clean WER | eval_other WER
--- | --- | --- | --- |--- |---
v1.7.0rc | SentencePiece Char | 6.94% | 7.38% | 19.49% | 22.73%

As far as I know, this is the best result that officially reported for Korean E2E ASR in ksponspeech data so far. As you mentioned, the performance of the model finetuned based on the pretrained encoder with the English was better than that trained from scratch.

**Reply by titu1994** [reply]
This is awesome ! These are greedy numbers ? I don't think there's baseline scores other than the paper, maybe I missed other frameworks. 

You could also add the score to papers with code if you like - https://paperswithcode.com/paper/kospeech-open-source-toolkit-for-end-to-end

**Reply by nycsv** [reply]
@titu1994 

Yeah I used greedy decoding. If it is used with other decoding method, the performance improves more. 

I wrote that this is the best result comparing [the recent paper](https://arxiv.org/abs/2110.05172) that reported performance based on Ksponspeech dataset.

**Reply by nycsv** [reply]
Hi, @duckyngo. Sorry, I don't have medium model.

**Reply by titu1994** [reply]
So reading table 2 and 3, the scores here with greedy decoding for clean sets is better than their "joint" method ? That's pretty impressive, and you didn't need SSL either so we could potentially further improve these scores. Really impressive !

**Reply by lifefeel** [reply]
Very awesome. Thank you for your great contribution. As a Korean, I also would like to participate your work. Send me DM, If you need help.

**Reply by nycsv** [reply]
@titu1994 That's right!  

I think that we can have better performance by applying other decoding methods or language model.

In addition, I have a plan to apply the Korean language to the SSL models such as wav2vec 2.0 and HuBERT. Are there any plans for development of the SSL models within Nemo? I'm currently experiencing that the SSL model within Nemo is unstable, right?

**Reply by titu1994** [reply]
@sam1373 has recently added quite good SSL support to Nemo, there's also a tutorial that will be released in Nemo 1.8 in the coming weeks. I think if currently supports contrastive learning based SSL but there are experiments ongoing to add HuBERT style as well.

**Comment by titu1994** [comment]
@eesungkim With the NeMo 1.8.1 release (soon™), we will support Huggingface Hub for external contributions (starting with ASR support). See https://github.com/NVIDIA/NeMo/pull/4030 for more details.

If you would like, you can upload a public checkpoint for Korean ASR to HuggingFace and add the links here so that others may use it easily.

When naming the model, please try to follow the current conversion for Conformer models- `stt_{2_char_lang_id}_conformer_{ctc/transducer}_{small/medium/large}`

So for example, `stt_kr_conformer_transducer_{small/medium}` would be the appropriate name for a 30 M param Conformer Transducer for the Korean language.

**Comment by okuchaiev** [comment]
@eesungkim It is now very easy to publish your model on Hugging Face Hub. @titu1994 prepared a great tutorial on how to do this https://github.com/NVIDIA/NeMo/discussions/4333   I would encourage you publish your model (under your name/org) on HF Hub.
